metadata-ingestion/src/datahub/ingestion/source/data_lake.py

```diff
@@ -0,0 +1,991 @@
+        analysis_metrics = AnalyzerContext.successMetricsAsJson(
+            self.spark, analysis_result
+        )
+
+        # reshape distinct counts into dictionary
+        column_distinct_counts = {
+            x["instance"]: int(x["value"])
+            for x in analysis_metrics
+            if x["name"] == "ApproxCountDistinct"
+        }
+
+        select_numeric_null_counts = [
+            count(
+                when(
+                    isnan(c) | col(c).isNull(),
+                    c,
+                )
+            ).alias(c)
+            for c in self.columns_to_profile
+            if column_types[column] in [DoubleType, FloatType]
+        ]
+
+        # PySpark doesn't support isnan() on non-float/double columns
+        select_nonnumeric_null_counts = [
+            count(
+                when(
+                    col(c).isNull(),
+                    c,
+                )
+            ).alias(c)
+            for c in self.columns_to_profile
+            if column_types[column] not in [DoubleType, FloatType]
+        ]
+
+        null_counts = dataframe.select(
+            select_numeric_null_counts + select_nonnumeric_null_counts
+        )
+        column_null_counts = null_counts.toPandas().T[0].to_dict()
+        column_null_fractions = {
+            c: column_null_counts[c] / self.row_count for c in self.columns_to_profile
+        }
+        column_nonnull_counts = {
+            c: self.row_count - column_null_counts[c] for c in self.columns_to_profile
+        }
+
+        column_unique_proportions = {
+            c: (
+                column_distinct_counts[c] / column_nonnull_counts[c]
+                if column_nonnull_counts[c] > 0
+                else 0
+            )
+            for c in self.columns_to_profile
+        }
+
+        if self.profiling_config.include_field_sample_values:
+            # take sample and convert to Pandas DataFrame
+            rdd_sample = dataframe.rdd.takeSample(False, NUM_SAMPLE_ROWS, seed=0)
+
+        # init column specs with profiles
+        for column in self.columns_to_profile:
+            column_profile = DatasetFieldProfileClass(fieldPath=column)
+
+            column_spec = _SingleColumnSpec(column, column_profile)
+
+            column_profile.uniqueCount = column_distinct_counts.get(column)
+            column_profile.uniqueProportion = column_unique_proportions.get(column)
+            column_profile.nullCount = column_null_counts.get(column)
+            column_profile.nullProportion = column_null_fractions.get(column)
+            if self.profiling_config.include_field_sample_values:
+                column_profile.sampleValues = [str(x[column]) for x in rdd_sample]
+
+            column_spec.type_ = column_types[column]
+            column_spec.cardinality = _convert_to_cardinality(
+                column_distinct_counts[column],
+                column_null_fractions[column],
+            )
+
+            self.column_specs.append(column_spec)
```

Comment:
what was the reason for splitting this into column spec + prepare table profile methods?

The SQL profiler did it for column-level parallel execution and query combining, but it doesn't seem that that's being used here?

Reply:
Deequ does some magic when you run multiple related analyzers on the same dataset so the number of scans is reduced (see [here](https://assets.amazon.science/65/22/91b90cfe4f85a6cc995e4d9d3f22/differential-data-quality-verification-for-partitioned-data.pdf) for details). That's why I decided to queue all of the analyzers into a single run and initiate the column specs beforehand so we can decide which ones to add.